okhttp/src/main/java/okhttp3/Body.java

```diff
@@ -0,0 +1,269 @@
+package okhttp3;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import okhttp3.internal.Util;
+import okhttp3.internal.http.HttpHeaders;
+import okio.Buffer;
+import okio.BufferedSink;
+import okio.BufferedSource;
+import okio.ByteString;
+import okio.Okio;
+import okio.Source;
+
+import static okhttp3.internal.Util.UTF_8;
+
+public abstract class Body implements Closeable {
+  public static final Body EMPTY = create(null, Util.EMPTY_BYTE_ARRAY);
+
+  private Reader charStreamReader;
+
+  @Override public void close() throws IOException {
+    Util.closeQuietly(source());
+  }
+
+  /**
+   * Returns the {@link MediaType} for this body.
+   */
+  public MediaType contentType() {
+    return null;
+  }
+
+  /**
+   * Returns the content length or -1 when the content length is unknown.
+   */
+  public long contentLength() throws IOException {
+    return -1L;
+  }
+
+  /**
+   * Returns a {@link BufferedSource} to read the contents of this {@link Body}.
+   * throws {link {@link UnsupportedOperationException}} if this body is not readable.
+   */
+  public BufferedSource source() {
+    throw new UnsupportedOperationException("Cannot read a non-readable body");
```

Comment:
Whenever you are extending Body most likely you only need to implement one of these 2 methods and the other will throw. We are just trying to save users of this class to have to implement a method that is not used.

Not really strong on this here since we have provided ReadableBody and WritableBody to facilitate that.